opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -50,424 +50,304 @@ impl<T: Number<T>> Aggregator<T> for ValueMap<T> {
+                dp.attributes = attrs;
+                dp.start_time = Some(prev_start);
+                dp.time = Some(t);
+                dp.value = value;
+                dp.exemplars.clear()
+            } else {
+                s_data.data_points.push(DataPoint {
+                    attributes: attrs,
+                    start_time: Some(prev_start),
+                    time: Some(t),
+                    value,
+                    exemplars: vec![],
+                });
+            }
+        }
         // The delta collection cycle resets.
         if let Ok(mut start) = self.start.lock() {
             *start = t;
         }
 
-        Some(Box::new(out))
+        (n, new_agg.map(|a| Box::new(a) as Box<_>))
     }
-}
-
-/// Returns an [Aggregator] that summarizes a set of measurements as their
-/// arithmetic sum.
-///
-/// Each sum is scoped by attributes and the aggregation cycle the measurements
-/// were made in.
-///
-/// The monotonic value is used to communicate the produced [Aggregation] is
-/// monotonic or not. The returned [Aggregator] does not make any guarantees this
-/// value is accurate. It is up to the caller to ensure it.
-///
-/// Each aggregation cycle is treated independently. When the returned
-/// Aggregator's Aggregation method is called it will reset all sums to zero.
-pub(crate) fn new_cumulative_sum<T: Number<T>>(monotonic: bool) -> Arc<dyn Aggregator<T>> {
-    Arc::new(CumulativeSum {
-        value_map: ValueMap::new(),
-        monotonic,
-        start: Mutex::new(SystemTime::now()),
-    })
-}
 
-/// Summarizes a set of measurements made over all aggregation cycles as their
-/// arithmetic sum.
-struct CumulativeSum<T: Number<T>> {
-    value_map: ValueMap<T>,
-    monotonic: bool,
-    start: Mutex<SystemTime>,
-}
+    pub(crate) fn cumulative(
+        &self,
+        dest: Option<&mut dyn Aggregation>,
+    ) -> (usize, Option<Box<dyn Aggregation>>) {
+        let t = SystemTime::now();
 
-impl<T: Number<T>> Aggregator<T> for CumulativeSum<T> {
-    fn aggregate(&self, measurement: T, attrs: AttributeSet) {
-        self.value_map.aggregate(measurement, attrs)
-    }
+        let s_data = dest.and_then(|d| d.as_mut().downcast_mut::<data::Sum<T>>());
+        let mut new_agg = if s_data.is_none() {
+            Some(data::Sum {
+                data_points: vec![],
+                temporality: Temporality::Cumulative,
+                is_monotonic: self.monotonic,
+            })
+        } else {
+            None
+        };
+        let s_data = s_data.unwrap_or_else(|| new_agg.as_mut().expect("present if s_data is none"));
+        s_data.temporality = Temporality::Cumulative;
+        s_data.is_monotonic = self.monotonic;
 
-    fn aggregation(&self) -> Option<Box<dyn Aggregation>> {
         let values = match self.value_map.values.lock() {
-            Ok(guard) => guard,
-            Err(_) => return None,
+            Ok(v) => v,
+            Err(_) => return (0, None),
```

Comment:
nit: The above feels like an open-coded [`Result::ok()`](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok)?

Reply:
It is a bit of that, but there are two separate cases as the first number represents the number of elements added to the aggregation or if it can be ignored (and the memory reused) for the 0 case, while the second is about if the previous data could be re-used in place vs creating a new aggregation to store the data for this cycle.